core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala

```diff
@@ -391,8 +398,10 @@ class UnifiedLogTest {
         assertNonEmptyFetch(log, offset, FetchIsolation.HIGH_WATERMARK)
       }
 
-      (log.highWatermark to log.logEndOffset).foreach { offset =>
-        assertEmptyFetch(log, offset, FetchIsolation.HIGH_WATERMARK)
+      assertEmptyFetch(log, log.highWatermark, FetchIsolation.HIGH_WATERMARK)
+
+      (log.highWatermark + 1 to log.logEndOffset).foreach { offset =>
```

Comment:
Why are we changing to start from log.highWatermark + 1 instead of log.highWatermark?

Reply:
since `log.highWatermark` contains the full-metadata, divided the check into 2 when reading at (or) beyond the high-watermark:

1) When the fetchOffset equals to the high-watermark, we return empty-records but with complete offset metadata
2) When the fetchOffset is beyond the high-watermark/max-offset-allowed-to-read, we return empty-records with message-only metadata. 

We can reconsider the case whether to return message-only-offset metadata (or) complete-offset metadata when the fetch-offset is beyond the `max-offset-allowed-to-read`.